File: 2016/day2.rs
use std::fs;

fn main() {
    part1();
    println!("\n");
    part2();
}

fn part1() {
    let mut nums = [1, 1];
    let data = fs::read_to_string("in.txt").unwrap();
    for i in data.split("\n"){
        for j in i.chars() {
            if j == 'D' && nums[1] < 2{
                nums[1] += 1
            } else if j == 'U' && nums[1] > 0 {
                nums[1] -= 1
            } else if j == 'R' && nums[0] < 2 {
                nums[0] += 1
            } else if j == 'L' && nums[0] > 0 {
                nums[0] -= 1
            }
        }
        println!("{}, {}", nums[0], nums[1])
    }
}

fn part2() {
    let mut nums = [0, 2];
    let data = fs::read_to_string("in.txt").unwrap();
    for i in data.split("\n"){
        for j in i.chars() {
            if j == 'D' && (nums[1] < 2 || (nums[1] < 3 && nums[0] != 0 && nums[0] != 4) || (nums[1] < 4 && nums[0] == 2)){
                nums[1] += 1
            } else if j == 'U' && (nums[1] > 2 || (nums[1] > 1 && nums[0] != 0 && nums[0] != 4) || (nums[1] > 0 && nums[0] == 2)) {
                nums[1] -= 1
            } else if j == 'R' && (nums[0] < 2 || (nums[0] < 3 && nums[1] != 0 && nums[1] != 4) || (nums[0] < 4 && nums[1] == 2)) {
                nums[0] += 1
            } else if j == 'L' && (nums[0] > 2 || (nums[0] > 1 && nums[1] != 0 && nums[1] != 4) || (nums[0] > 0 && nums[1] == 2)) {
                nums[0] -= 1
            }
        }
        println!("{}, {}", nums[0], nums[1])
    }
}

File: 2017/day2.ts
import * as fs from "fs";

function part1(): number {
    var data = fs.readFileSync("in.txt", "utf8").split("\r\n")
    var res = 0
    for (let i = 0; i < data.length; i++) {
        var nums = data[i].split("\t")
        var maximum = 0
        var minimum = 9999
        for (let j = 0; j < nums.length; j++) {
            if (Number(nums[j]) > maximum) {
                maximum = Number(nums[j])
            }
            if (Number(nums[j]) < minimum) {
                minimum = Number(nums[j])
            }
        }
        res += maximum - minimum
    }
    return res
}

function part2(): number {
    var data = fs.readFileSync("in.txt", "utf8").split("\r\n")
    var res = 0
    for (let i = 0; i < data.length; i++) {
        var nums = data[i].split("\t")
        for (let j = 0; j < nums.length; j++) {
            for (let k = 0; k < nums.length; k++) {
                if ((k != j) && (Number(nums[k]) >= Number(nums[j])) && ((nums[k] % nums[j]) == 0)) {
                    res += nums[k] / nums[j]
                }
            }
        }
    }
    return res
}

console.log("Part 1:", part1())
console.log("Part 1:", part2())